proto/stream_plan.proto

```diff
@@ -496,6 +496,11 @@ message ChainNode {
 
   // Snapshot read every N barriers
   uint32 snapshot_read_barrier_interval = 9 [deprecated = true];
+
+  // backfill snapshot read, only used in madsim tests
+  // Otherwise snapshot read will complete almost instantaneously,
+  // and the backfill can't test upstream.
+  uint32 snapshot_read_delay = 10;
```

Comment:
This looks dirty to me. 😕 Can we simply leverage the rate limit here?

Reply:
Tried it previously. It could not help in this case, although I'm not sure why.
(It's not because it's disabled in madsim, I added support for it in madsim as well, but does not resolve it still).

Reply:
The stream job will still complete, before `cancel` can happen

Reply:
cc @wangrunji0408 Do you have better ideas? Does `time::advance` help here?

Reply:
> cc @wangrunji0408 Do you have better ideas? Does `time::advance` help here?

`time::advance` only advances the actual time. What we want here is to yield to the tokio runtime, so other threads of execution can run. Hence we use sleep. In this case we want to make sure to add sufficient delay, such that `meta` is able to execute the cancel.

Reply:
I will use an env var for this first. It will just be a one line change, and easy to revert if we discover the reason why `rate limit` does not work.

Reply:
No idea about why it didn't work. But you are always free to insert `#[cfg(madsim)]` code to add test-specific logic. +1 for not exposing user-facing configs.

Reply:
> `time::advance` only advances the actual time. What we want here is to yield to the tokio runtime, so other threads of execution can run.

You're right. What I want to say might be `task::consume_budget` or `task::yield_now`.

https://github.com/risingwavelabs/risingwave/blob/e392db0c6f468caf89597294a94d4e4134c61eaf/src/storage/src/hummock/sstable/forward_sstable_iterator.rs#L245-L248

Reply:
> I will use an env var for this first. It will just be a one line change, and easy to revert if we discover the reason why rate limit does not work.

Resolved https://github.com/risingwavelabs/risingwave/pull/13070/commits/5bbb840a60e7e77d9eb82fa5d0bbf4b5b658f283, https://github.com/risingwavelabs/risingwave/pull/13070/commits/28b19272b5b3f5e377f1b5b41650fa15e325f6fa